README.md
![go-build](https://github.com/ChrIgiSta/stopwatch/actions/workflows/build.yml/badge.svg) ![go-test](https://github.com/ChrIgiSta/stopwatch/actions/workflows/ci.yml/badge.svg)

# App 
This app enables stopwatch functionality for eliona.

![SmartView Picture not fould](eliona/example_stopwatch_view.png?raw=true "SmartView with stopwatch")

## Configuration

This app needs no configuration over api or database. Just use the new added asset type `Stopwatch` to start and stop an second counter as many you want.

### Registration in Eliona ###

To start and initialize an app in an Eliona environment, the app have to registered in Eliona. For this, an entry `stopwatch` in the database table `public.eliona_app` is necessary.

### Environment variables

- `APPNAME`: must be set to `stopwatch`. Some resources use this name to identify the app inside an Eliona environment.

- `CONNECTION_STRING`: configures the [Eliona database](https://github.com/eliona-smart-building-assistant/go-eliona/tree/main/db). Otherwise, the app can't be initialized and started. (e.g. `postgres://user:pass@localhost:5432/iot`)

- `API_ENDPOINT`:  configures the endpoint to access the [Eliona API v2](https://github.com/eliona-smart-building-assistant/eliona-api). Otherwise, the app can't be initialized and started. (e.g. `http://api-v2:3000/v2`)

- `API_TOKEN`: defines the secret to authenticate the app and access the API. 

- `API_SERVER_PORT`(optional): define the port the API server listens. The default value is Port `3000`.

- `DEBUG_LEVEL`(optional): defines the minimum level that should be [logged](https://github.com/eliona-smart-building-assistant/go-eliona/tree/main/log). Not defined the default level is `info`.

### Database tables ###

This app works without any database tables.

### App API ###

The app provides its own API to access configuration data and other functions. The full description of the API is defined in the `openapi.yaml` OpenAPI definition file.

- [API Reference](https://eliona-smart-building-assistant.github.io/open-api-docs/?https://raw.githubusercontent.com/eliona-smart-building-assistant/stopwatch/develop/openapi.yaml) shows Details of the API

**Generation**: to generate api server stub see Generation section below.


### Eliona ###

This app provides a start and a stop input (Button) to start or stop the stopwatch. In Current Value the time in seconds will provided to eliona until the timer has stopped. Then the Last Value will be actualized and the Current Value will be set to 0.

## Tools

### Testing ###

The code can be tested by running `go test`

### Generate API server stub ###

For the API server the [OpenAPI Generator](https://openapi-generator.tech/docs/generators/openapi-yaml) for go-server is used to generate a server stub. The easiest way to generate the server files is to use one of the predefined generation script which use the OpenAPI Generator Docker image.

```
.\generate-api-server.cmd # Windows
./generate-api-server.sh # Linux
```

### Generate Database access ###

For the database access [SQLBoiler](https://github.com/volatiletech/sqlboiler) is used. The easiest way to generate the database files is to use one of the predefined generation script which use the SQLBoiler implementation. Please note that the database connection in the `sqlboiler.toml` file have to be configured.

```
.\generate-db.cmd # Windows
./generate-db.sh # Linux
```

## Build on your own

 * Native
    * Install deps: `go install .`
    * Build: `go build .`
 * Docker
    * `docker build . --tag my-stopwatch`


CHANGELOG.md
# Changelog

This file contains the repo-tags with the corresponding information.


## [v1.0.1] - 2022-10-24

- graceful shutdown at os signals
- change subtype for attribute `last_time` to `info`
- add a CHANGELOG

## [v1.0.0] - 2022-10-21 :partying_face:

- new eliona app to bring eliona stopwatch functionality
- first release


eliona/data.go
//  This file is part of the eliona project.
//  Copyright © 2022 LEICOM iTEC AG. All Rights Reserved.
//  ______ _ _
// |  ____| (_)
// | |__  | |_  ___  _ __   __ _
// |  __| | | |/ _ \| '_ \ / _` |
// | |____| | | (_) | | | | (_| |
// |______|_|_|\___/|_| |_|\__,_|
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
//  BUT NOT LIMITED  TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NON INFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
//  DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

package eliona

import (
	"context"
	"encoding/json"
	"stopwatch/conn"

	api "github.com/eliona-smart-building-assistant/go-eliona-api-client/v2"
	"github.com/eliona-smart-building-assistant/go-eliona/client"
	"github.com/eliona-smart-building-assistant/go-utils/log"
)

const (
	ATTRIBUTE_START        = "start"
	ATTRIBUTE_STOP         = "stop"
	ATTRIBUTE_CURRENT_TIME = "current_time"
	ATTRIBUTE_LAST_TIME    = "last_time"
	SUBTYPE_ACTIONS        = api.SUBTYPE_OUTPUT
	SUBTYPE_VALUES         = api.SUBTYPE_INPUT
	SUBTYPE_LASTTIME       = api.SUBTYPE_INFO
	RX_BUFFER              = 20000
	WEBSOCKET_ENDPOINT     = "/data-listener"
)

func UpdateTime(assetId int32, newTime float64) error {

	dataData := make(map[string]interface{})
	dataData[ATTRIBUTE_CURRENT_TIME] = newTime

	data := api.Data{
		AssetId: assetId,
		Subtype: SUBTYPE_VALUES,
		// Timestamp:     *api.NewNullableTime(nil),
		Data: dataData,
	}
	resp, err := client.NewClient().DataApi.PutData(context.Background()).Data(data).Execute()

	log.Debug(MODULE, "get current time response: %v", resp)

	return err
}

func UpdateLastTime(assetId int32, newTime float64) error {

	dataData := make(map[string]interface{})
	dataData[ATTRIBUTE_LAST_TIME] = newTime

	data := api.Data{
		AssetId: assetId,
		Subtype: SUBTYPE_LASTTIME,
		// Timestamp:     *api.NewNullableTime(nil),
		Data: dataData,
	}
	resp, err := client.NewClient().DataApi.PutData(context.Background()).Data(data).Execute()

	log.Debug(MODULE, "get last time response: %v", resp)

	return err
}

func ListenHeapEvents(ir <-chan bool, rxApiData chan<- api.Data) {
	var (
		rx chan []byte
	)
	defer close(rxApiData)

	rx = make(chan []byte, RX_BUFFER)

	ws := conn.NewWebsocketClient(client.ApiEndpointString()+WEBSOCKET_ENDPOINT, false)
	go ws.ServeForever(rx, ir)
	for data := range rx {
		var apiData api.Data
		err := json.Unmarshal(data, &apiData)
		if err == nil {
			// prefilter data
			if apiData.Subtype == SUBTYPE_ACTIONS {
				rxApiData <- apiData
			}
		} else {
			log.Warn(MODULE, "cannot unmarshal ws data", err)
		}
	}
	log.Warn(MODULE, "rx channel from ws closed. exiting event listener...")
}


app.go
//  This file is part of the eliona project.
//  Copyright © 2022 LEICOM iTEC AG. All Rights Reserved.
//  ______ _ _
// |  ____| (_)
// | |__  | |_  ___  _ __   __ _
// |  __| | | |/ _ \| '_ \ / _` |
// | |____| | | (_) | | | | (_| |
// |______|_|_|\___/|_| |_|\__,_|
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
//  BUT NOT LIMITED  TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NON INFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
//  DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

package main

import (
	"context"
	"net/http"
	"os"
	"os/signal"
	"stopwatch/apiserver"
	"stopwatch/apiservices"
	"stopwatch/eliona"
	"stopwatch/stopwtch"
	"syscall"
	"time"

	api "github.com/eliona-smart-building-assistant/go-eliona-api-client/v2"
	"github.com/eliona-smart-building-assistant/go-utils/common"
	"github.com/eliona-smart-building-assistant/go-utils/log"
)

const (
	MODULE          = "app"
	API_DATA_BUFFER = 20000
)

var (
	stopwatches []api.Asset
	ir          chan bool
	swMng       stopwtch.StopwatchManager
	apiServer   *http.Server
	osIr        bool
)

func actualizeStopwatches() {
	var err error

	stopwatches, err = eliona.GetStopwatches()
	log.Debug(MODULE, "got stopwatches %v with err %v", stopwatches, err)
}

func setupApp() {
	osIr = false

	log.Info(MODULE, "setup application")

	var (
		apiData chan api.Data
	)

	ir = make(chan bool)
	apiData = make(chan api.Data, API_DATA_BUFFER)

	actualizeStopwatches()

	swMng = *stopwtch.NewStopwatchManager(stopwatchCallbackHandler)

	go eliona.ListenHeapEvents(ir, apiData)
	go stopwatchEventCatcher(apiData)
	go catchAndExitOnOsSig()
}

func catchAndExitOnOsSig() {
	osSig := make(chan os.Signal, 1)
	signal.Notify(osSig,
		syscall.SIGHUP,
		syscall.SIGINT,
		syscall.SIGTERM,
		syscall.SIGQUIT)

	<-osSig
	osIr = true
	log.Info("app", "Os Signal catched, exiting...")
	ir <- true
	swMng.StopAll()
	log.Info("app", "all stopwatches stopped")
	apiServer.Shutdown(context.Background())
	log.Info("app", "api server stopped")
}

// start / stop stopwatches
func stopwatchEventCatcher(apiData <-chan api.Data) {
	for data := range apiData {
		log.Debug(MODULE, "data from datalistener: %v", data)

		asset := getStopwatchByAssetId(data.AssetId)
		if asset != nil {
			processStartStopEvent(data.AssetId, data.Data)
		}
	}
	log.Warn(MODULE, "eventcatcher exited")
}

func processStartStopEvent(assetId int32, data map[string]interface{}) {
	start := data[eliona.ATTRIBUTE_START]
	stop := data[eliona.ATTRIBUTE_STOP]

	// prioritize stop signal
	if stop != nil && stop.(float64) >= 1 {
		lastTime := swMng.Stop(assetId)
		if lastTime.Seconds() > 0 {
			log.Info(MODULE, "timer stoppend %d @ %f s",
				assetId,
				lastTime.Seconds())
			eliona.UpdateLastTime(assetId, lastTime.Seconds())
		}
	} else if start != nil && start.(float64) >= 1 {
		log.Info(MODULE, "timer started %d", assetId)
		swMng.Start(assetId)
	}
}

// stopwatch ticks
func stopwatchCallbackHandler(assetId int32, t time.Duration) {
	log.Debug(MODULE, "stopwatch callback called %d,%v", assetId, t)

	eliona.UpdateTime(assetId, t.Seconds())
}

func getStopwatchByAssetId(assetId int32) *api.Asset {
	for _, sw := range stopwatches {
		if sw.GetId() == assetId {
			return &sw
		}
	}
	return nil
}

// listenApiRequests starts an API server and listen for API requests
// The API endpoints are defined in the openapi.yaml file
func listenApiRequests() {
	apiServer = &http.Server{
		Addr: ":" + common.Getenv("API_SERVER_PORT", "3000"),
		Handler: apiserver.NewRouter(
			apiserver.NewUtilsApiController(apiservices.NewUtilsApiService()),
		),
	}

	err := apiServer.ListenAndServe()

	if !osIr {
		log.Fatal(MODULE, "Error in API Server: %v", err)
	} else {
		log.Info(MODULE, "Api Server stopped todue an os interrupt")
	}
}
